### PR TITLE
Support index feeds

### PR DESCRIPTION
**Pull request by staltz**
- [x] Pending on https://github.com/ssb-ngi-pointer/ssb-index-feed-writer/pull/12

### CODE REVIEW

File: test/index-feeds.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      outputDir: 'ssb-fixtures-test-index-feeds',
+      seed: 'banana',
+      messages: M,
+      authors: A,
+      indexFeeds: I,
+    },
+    async (err, msgs, cleanup, outputDir) => {
+      // (seed + announce + add main + add indexes + add 5 index) * A
+      const META_MSG_COUNT = 9 * A;
+      // Seems like we should have only `M` index feed msgs, right?
```
Comment (mycognosist) [inline test/index-feeds.js:24]: Good call including these clarifying comments!

File: src/bin.ts [hunk 2]
```diff
@@ -33,6 +33,11 @@ const argv = yargs(process.argv)
     describe: 'Output follow-graph.json too',
   })
 
+  .option('indexFeeds', {
+    default: defaults.INDEX_FEEDS,
+    describe: 'Percentage (0–100) of authors to write index feeds',
```
Comment (mycognosist) [inline src/bin.ts:38]: It's awesome to have this level of control over the output.

**Review by mycognosist — APPROVED** [review]
I can't find any obvious holes in this PR. Compact code, as always.

I just have one clarifying question in the comments (related to `.some()` usage).

File: src/index-feeds.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+import {paretoSample} from './sample';
+
+function sampleAuthors(seed: string, authors: Array<Author>, total: number) {
+  const sampled: Array<FeedId> = [];
+  // Sample other authors, but don't sample ones that are already recipient
+  while (sampled.length < total) {
+    let other: FeedId;
+    do {
+      other = paretoSample(seed, authors).id;
+    } while (sampled.some((r) => other === r));
```
[the remark this answers is not in the document]
Comment (staltz) [reply src/index-feeds.ts:28]: Good to clarify! It's a bit confusing procedure indeed, but I copied this one from `generate.ts` file, so it should be working. It should read like this:

```
do {
  sample a new `other`
} while (`other` is something we already have sampled)
```

Basically meaning "do it over and over again until the `while` condition is false"

File: test/index-feeds.js (hunk 1 above)
Comment (staltz) [reply test/index-feeds.js:24]: This looks simple but it was torturing my mind to discover it :D

File: src/index-feeds.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mycognosist) [reply src/index-feeds.ts:28]: Ah OK, that makes more sense! "Keep generating a new `other` until `sampled` is filled-up with unique ids".

Thanks for clarifying.

File: test/index-feeds.js (hunk 1 above)
Comment (mycognosist) [reply test/index-feeds.js:24]: Haha I could tell it must have been a mind-bender if you took the time to document it like this :D Well done.

**Comment by mycognosist** [comment]
🟢 🙌🏻

File: src/index-feeds.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (staltz) [reply src/index-feeds.ts:28]: I'll add your quoted text as a code comment :)

**Pull request merged**